src/editor-service/ast-from-source-position.ts

```diff
@@ -0,0 +1,280 @@
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as parse5 from 'parse5';
+
+import {ParsedHtmlDocument} from '../html/html-document';
+import {SourceRange} from '../model/model';
+
+import {SourcePosition} from './editor-service';
+
+
+export type LocationResult = LocatedAttribute | LocatedAttributeValue |
+    LocatedTag | LocatedEndTag | LocatedInText | LocatedInScript |
+    LocatedInStyle | LocatedInComment;
+/** In the tagname of a start tag. */
+export interface LocatedTag {
+  kind: 'tagName';
+  element: parse5.ASTNode;
+}
+/** In an end tag. */
+export interface LocatedEndTag {
+  kind: 'endTag';
+  element: parse5.ASTNode;
+}
+/** In the attributes section of a start tag. Maybe in an attribute name. */
+export interface LocatedAttribute {
+  kind: 'attribute';
+  attribute: string|null;
+  element: parse5.ASTNode;
+}
+/** In the value of an attribute of a start tag. */
+export interface LocatedAttributeValue {
+  kind: 'attributeValue';
+  attribute: string;
+  element: parse5.ASTNode;
+}
+/** In a text node. */
+export interface LocatedInText {
+  kind: 'text';
+  textNode?: parse5.ASTNode;
+}
+/** In the text of a <script> */
+export interface LocatedInScript {
+  kind: 'scriptTagContents';
+  textNode?: parse5.ASTNode;
+}
+/** In the text of a <style> */
+export interface LocatedInStyle {
+  kind: 'styleTagContents';
+  textNode?: parse5.ASTNode;
+}
+/** In a comment. */
+export interface LocatedInComment {
+  kind: 'comment';
+  commentNode: parse5.ASTNode;
+}
+
+/**
+ * Given a position and an HTML document, try to describe what new text typed
+ * at the given position would be.
+ *
+ * Where possible we try to return the ASTNode describing that position, but
+ * sometimes there does not actually exist one. (for a simple case, the empty
+ * string should be interpreted as a text node, but there is no text node in
+ * an empty document, but there would be after the first character was typed).
+ */
+export function getLocationInfoForPosition(
+    document: ParsedHtmlDocument, position: SourcePosition): LocationResult {
+  const location =
+      _getLocationInfoForPosition(document.ast, position, document);
+  if (!location) {
+    return {kind: 'text'};
+  }
+  return location;
+}
+function _getLocationInfoForPosition(
+    node: parse5.ASTNode, position: SourcePosition,
+    document: ParsedHtmlDocument): undefined|LocationResult {
+  const sourceRange = document.sourceRangeForNode(node);
+  const location = node.__location;
+  if (!(sourceRange && location)) {
```

Comment:
can you document the reasoning behind this condition? it seems to be that if you can't get a SourceRange for a node, you recurse into its children, but how would you know that's the right thing to do if you can't get a SourceRange? Naively I would think this case should return `null`.